File: backend/ai_engine/backtester.py
import numpy as np
import pandas as pd
from datetime import datetime

def run_strategy_backtest(df_price_history, initial_capital=100000.0, min_confidence=60.0):
    """
    Backtests AI Buy Signal Strategy vs Buy & Hold baseline strategy.
    Expects df_price_history with columns: date, close, rsi, macd_hist, etc.
    """
    if len(df_price_history) < 20:
        # Generate clean synthetic backtest series if historical length is limited
        dates = pd.date_range('2023-01-01', '2026-06-30', freq='B')
        n = len(dates)
        price = 200.0
        prices = []
        for _ in range(n):
            price *= (1 + np.random.normal(0.0004, 0.012))
            prices.append(price)
            
        df_price_history = pd.DataFrame({'date': dates.strftime('%Y-%m-%d'), 'close': prices})
        df_price_history['rsi'] = 50 + np.sin(np.linspace(0, 20, n)) * 20
        df_price_history['macd_hist'] = np.cos(np.linspace(0, 15, n)) * 2.0
    
    df = df_price_history.copy().sort_values('date').reset_index(drop=True)
    n = len(df)
    
    # 1. Buy & Hold Strategy Calculation
    start_price = df['close'].iloc[0]
    bh_shares = initial_capital / start_price
    bh_equity = df['close'] * bh_shares
    
    # 2. AI Signal Strategy Simulation
    ai_equity = []
    ai_cash = initial_capital
    ai_shares = 0
    in_position = False
    trades = 0
    winning_trades = 0
    entry_price = 0.0
    
    for i in range(n):
        curr_price = df['close'].iloc[i]
        rsi = df.get('rsi', pd.Series([50]*n)).iloc[i]
        macd_h = df.get('macd_hist', pd.Series([0]*n)).iloc[i]
        
        # Simulated AI Signal rule: RSI recovery + positive MACD histogram
        is_ai_buy = (rsi < 62.0 and rsi > 45.0 and macd_h > 0.1)
        is_ai_sell = (rsi > 70.0 or macd_h < -0.3)
        
        if not in_position and is_ai_buy:
            # Enter Long Position
            ai_shares = ai_cash / curr_price
            ai_cash = 0
            in_position = True
            entry_price = curr_price
            trades += 1
        elif in_position and is_ai_sell:
            # Exit Position
            ai_cash = ai_shares * curr_price
            if curr_price > entry_price:
                winning_trades += 1
            ai_shares = 0
            in_position = False
            
        current_total_val = ai_cash + (ai_shares * curr_price)
        ai_equity.append(current_total_val)
        
    df['ai_equity'] = ai_equity
    df['bh_equity'] = bh_equity
    
    # Calculate performance stats
    final_ai_val = ai_equity[-1]
    final_bh_val = bh_equity.iloc[-1]
    
    ai_return = ((final_ai_val - initial_capital) / initial_capital) * 100.0
    bh_return = ((final_bh_val - initial_capital) / initial_capital) * 100.0
    
    years = max(1.0, n / 252.0)
    ai_cagr = (((final_ai_val / initial_capital) ** (1.0 / years)) - 1.0) * 100.0
    bh_cagr = (((final_bh_val / initial_capital) ** (1.0 / years)) - 1.0) * 100.0
    
    # Drawdowns
    ai_series = pd.Series(ai_equity)
    ai_peak = ai_series.cummax()
    ai_dd = ((ai_series - ai_peak) / ai_peak) * 100.0
    max_ai_dd = abs(ai_dd.min())
    
    bh_series = pd.Series(bh_equity)
    bh_peak = bh_series.cummax()
    bh_dd = ((bh_series - bh_peak) / bh_peak) * 100.0
    max_bh_dd = abs(bh_dd.min())
    
    # Sharpe Ratio (assumes 5% risk free rate)
    returns = ai_series.pct_change().dropna()
    sharpe = round((returns.mean() * 252 - 0.05) / (returns.std() * np.sqrt(252) + 1e-9), 2)
    win_rate = round((winning_trades / max(1, trades)) * 100.0, 1)

    chart_data = []
    # Sample down chart points if length > 150 for crisp visualization
    step = max(1, n // 120)
    for i in range(0, n, step):
        chart_data.append({
            "date": str(df['date'].iloc[i])[:10],
            "aiStrategy": round(float(df['ai_equity'].iloc[i]), 2),
            "buyAndHold": round(float(df['bh_equity'].iloc[i]), 2),
            "price": round(float(df['close'].iloc[i]), 2)
        })

    return {
        "summary": {
            "initialCapital": initial_capital,
            "aiFinalValue": round(final_ai_val, 2),
            "bhFinalValue": round(final_bh_val, 2),
            "aiReturnPct": round(ai_return, 2),
            "bhReturnPct": round(bh_return, 2),
            "aiCagr": round(ai_cagr, 2),
            "bhCagr": round(bh_cagr, 2),
            "maxAiDrawdown": round(max_ai_dd, 2),
            "maxBhDrawdown": round(max_bh_dd, 2),
            "sharpeRatio": sharpe,
            "totalTrades": trades,
            "winRate": win_rate
        },
        "equityCurve": chart_data
    }

File: backend/src/ai/backtester.js
// JS port of ai_engine/backtester.py — AI Buy Signal Strategy vs Buy & Hold.
// Input: ascending OHLCV rows. Output: same shape the frontend expects.

const { computeAllIndicators } = require('./indicators');

const round2 = (v) => Math.round(v * 100) / 100;

function runStrategyBacktest(rows, initialCapital = 100000, minConfidence = 60) {
  const ind = computeAllIndicators(rows);
  const n = ind.length;

  if (n < 20) {
    return {
      error: true,
      message: `Insufficient price history for backtesting. Need at least 20 data points, got ${n}.`,
      summary: {
        initialCapital,
        aiFinalValue: initialCapital,
        bhFinalValue: initialCapital,
        aiReturnPct: 0.0,
        bhReturnPct: 0.0,
        aiCagr: 0.0,
        bhCagr: 0.0,
        maxAiDrawdown: 0.0,
        maxBhDrawdown: 0.0,
        sharpeRatio: 0.0,
        totalTrades: 0,
        winRate: 0.0,
      },
      equityCurve: [],
    };
  }

  const sorted = ind.slice().sort((a, b) => (a.date < b.date ? -1 : a.date > b.date ? 1 : 0));

  // 1. Buy & Hold strategy
  const startPrice = sorted[0].close;
  const bhShares = initialCapital / startPrice;
  const bhEquity = sorted.map(r => r.close * bhShares);

  // 2. AI signal simulation
  const aiEquity = [];
  let aiCash = initialCapital;
  let aiShares = 0;
  let inPosition = false;
  let trades = 0;
  let winningTrades = 0;
  let entryPrice = 0.0;

  for (let i = 0; i < n; i++) {
    const currPrice = sorted[i].close;
    const rsi = sorted[i].rsi != null ? sorted[i].rsi : 50;
    const macdH = sorted[i].macd_hist != null ? sorted[i].macd_hist : 0;

    const isAiBuy = rsi < 62.0 && rsi > 45.0 && macdH > 0.1;
    const isAiSell = rsi > 70.0 || macdH < -0.3;

    if (!inPosition && isAiBuy) {
      aiShares = aiCash / currPrice;
      aiCash = 0;
      inPosition = true;
      entryPrice = currPrice;
      trades += 1;
    } else if (inPosition && isAiSell) {
      aiCash = aiShares * currPrice;
      if (currPrice > entryPrice) winningTrades += 1;
      aiShares = 0;
      inPosition = false;
    }

    aiEquity.push(aiCash + aiShares * currPrice);
  }

  const finalAiVal = aiEquity[n - 1];
  const finalBhVal = bhEquity[n - 1];

  const aiReturn = ((finalAiVal - initialCapital) / initialCapital) * 100.0;
  const bhReturn = ((finalBhVal - initialCapital) / initialCapital) * 100.0;

  const years = Math.max(1.0, n / 252.0);
  const aiCagr = (Math.pow(finalAiVal / initialCapital, 1.0 / years) - 1.0) * 100.0;
  const bhCagr = (Math.pow(finalBhVal / initialCapital, 1.0 / years) - 1.0) * 100.0;

  // Drawdowns
  let aiPeak = -Infinity;
  let maxAiDd = 0;
  let bhPeak = -Infinity;
  let maxBhDd = 0;
  for (let i = 0; i < n; i++) {
    if (aiEquity[i] > aiPeak) aiPeak = aiEquity[i];
    if (aiPeak > 0) maxAiDd = Math.max(maxAiDd, Math.abs(((aiEquity[i] - aiPeak) / aiPeak) * 100));
    if (bhEquity[i] > bhPeak) bhPeak = bhEquity[i];
    if (bhPeak > 0) maxBhDd = Math.max(maxBhDd, Math.abs(((bhEquity[i] - bhPeak) / bhPeak) * 100));
  }

  // Sharpe ratio (assumes 5% risk-free rate; sample std like pandas ddof=1)
  const returns = [];
  for (let i = 1; i < n; i++) returns.push(aiEquity[i] / aiEquity[i - 1] - 1);
  let sharpe = 0;
  if (returns.length >= 2) {
    const mean = returns.reduce((s, v) => s + v, 0) / returns.length;
    const variance = returns.reduce((s, v) => s + (v - mean) * (v - mean), 0) / (returns.length - 1);
    const std = Math.sqrt(variance);
    if (isFinite(std)) sharpe = (mean * 252 - 0.05) / (std * Math.sqrt(252) + 1e-9);
  }
  sharpe = round2(sharpe);

  const winRate = Math.round((winningTrades / Math.max(1, trades)) * 100 * 10) / 10;

  const chartData = [];
  const step = Math.max(1, Math.floor(n / 120));
  for (let i = 0; i < n; i += step) {
    chartData.push({
      date: String(sorted[i].date).slice(0, 10),
      aiStrategy: round2(aiEquity[i]),
      buyAndHold: round2(bhEquity[i]),
      price: round2(sorted[i].close),
    });
  }

  return {
    summary: {
      initialCapital,
      aiFinalValue: round2(finalAiVal),
      bhFinalValue: round2(finalBhVal),
      aiReturnPct: round2(aiReturn),
      bhReturnPct: round2(bhReturn),
      aiCagr: round2(aiCagr),
      bhCagr: round2(bhCagr),
      maxAiDrawdown: round2(maxAiDd),
      maxBhDrawdown: round2(maxBhDd),
      sharpeRatio: sharpe,
      totalTrades: trades,
      winRate,
    },
    equityCurve: chartData,
  };
}

module.exports = { runStrategyBacktest };
